File: interfaces/python/tests/unittests/unconstrained_tests.py
# Test osqp python module
import osqp
# import osqppurepy as osqp
import numpy as np
from scipy import sparse
import scipy as sp

# Unit Test
import unittest
import numpy.testing as nptest


class unconstrained_tests(unittest.TestCase):

    def setUp(self):
        """
        Setup unconstrained quadratic problem
        """
        # Unconstrained QP problem
        sp.random.seed(4)

        self.n = 30
        self.m = 0
        P = sparse.diags(np.random.rand(self.n)) + 0.2*sparse.eye(self.n)
        self.P = P.tocsc()
        self.q = np.random.randn(self.n)
        self.A = sparse.csc_matrix((self.m, self.n))
        self.l = np.array([])
        self.u = np.array([])
        self.opts = {'verbose': False,
                     'eps_abs': 1e-08,
                     'eps_rel': 1e-08,
                     'polish': False}
        self.model = osqp.OSQP()
        self.model.setup(P=self.P, q=self.q, A=self.A, l=self.l, u=self.u,
                         **self.opts)

    def test_unconstrained_problem(self):

        # Solve problem
        res = self.model.solve()

        # Assert close
        nptest.assert_array_almost_equal(
            res.x, np.array([
                -0.61981415, -0.06174194, 0.83824061, -0.0595013, -0.17810828,
                2.90550031, -1.8901713, -1.91191741, -3.73603446, 1.7530356,
                -1.67018181, 3.42221944, 0.61263403, -0.45838347, -0.13194248,
                2.95744794, 5.2902277, -1.42836238, -8.55123842, -0.79093815,
                0.43418189, -0.69323554, 1.15967924, -0.47821898, 3.6108927,
                0.03404309, 0.16322926, -2.17974795, 0.32458796, -1.97553574]))
        nptest.assert_array_almost_equal(res.y, np.array([]))
        nptest.assert_array_almost_equal(res.info.obj_val, -35.020288603855825)

File: interfaces/matlab/unittests/unconstrained_tests.m
classdef unconstrained_tests < matlab.unittest.TestCase
    %UNCONSTRAINED_TESTS Solve unconstrained quadratic program
    
    properties
        P
        q
        A
        u
        l
        m,
        n
        solver
        options
        tol
    end
    
    methods(TestMethodSetup)
        function setup_problem(testCase)
            % Create Problem
            rng(4)
            testCase.n = 30;
            testCase.m = 0;
            testCase.P = sparse(diag(rand(testCase.n, 1))) + 0.2*speye(testCase.n);
            testCase.q = randn(testCase.n, 1);
            testCase.A = [];
            testCase.l = [];
            testCase.u = [];
            
            
            % Setup solver
            testCase.solver = osqp;
            testCase.solver.setup(testCase.P, testCase.q, ...
                testCase.A, testCase.l, testCase.u, 'verbose', 0, 'eps_abs', 1e-05, 'eps_rel', 1e-05);
            
            % Get options
            testCase.options = testCase.solver.current_settings();
            
            % Setup tolerance
            testCase.tol = 1e-04;
            
        end
    end
    
    methods (Test)
        function test_unconstrained_problem(testCase)
            % Solve with OSQP
            results = testCase.solver.solve();
            
            % Check if they are close
            testCase.verifyEqual(results.x, ...
                [-0.1139; -2.6464; -0.0414; 0.4539; 0.8058; -0.2232; -0.0399; -1.6256; 0.4702; ...
                 -0.6265; 0.3740; -0.6547; -0.4346; 0.9082; 0.8801; -0.0094; -2.3109; 0.5990; ...
                 -0.4948; -0.1263; -3.3029; 0.2563; -0.6106; 0.4830; 0.0081; 4.3573; -2.9165; ...
                 0.4514; -1.7058; 0.5228], 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.y, zeros(0,1), 'AbsTol',testCase.tol)
            testCase.verifyEqual(results.info.obj_val, -16.3649, 'AbsTol', testCase.tol)
            
        end
        
    end
    
end
